soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,355 @@
+
+    /// Hashes a message `msg` to a point in G1, using a domain separation tag `dst`.
+    pub fn hash_to_g1(&self, msg: &Bytes, dst: &Bytes) -> G1Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_hash_to_g1(env, msg.into(), dst.to_object())
+            .unwrap_infallible();
+        unsafe { G1Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) }
+    }
+
+    // g2
+
+    /// Checks if a point `p` in G2 is in the correct subgroup.
+    pub fn g2_is_in_subgroup(&self, p: &G2Affine) -> bool {
+        let env = self.env();
+        let res = internal::Env::bls12_381_check_g2_is_in_subgroup(env, p.to_object())
+            .unwrap_infallible();
+        res.into()
+    }
+
+    /// Adds two points `p0` and `p1` in G2.
+    pub fn g2_add(&self, p0: &G2Affine, p1: &G2Affine) -> G2Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g2_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        unsafe { G2Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) }
+    }
+
+    /// Adds two points `p0` and `p1` in G2, ensuring that the result is in the
+    /// correct subgroup. Note the subgroup check is computationally expensive,
+    /// so if want to perform a series of additions i.e. `agg = p0 + p1 + .. +pn`,     
+    /// it may make sense to only call g2_checked_add on the final addition,
+    /// while using `g2_add` (non-checked version) on the intermediate ones.
+    pub fn g2_checked_add(&self, p0: &G2Affine, p1: &G2Affine) -> Option<G2Affine> {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g2_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        let res = unsafe { G2Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) };
+        let is_in_correct_subgroup: bool =
+            internal::Env::bls12_381_check_g2_is_in_subgroup(env, res.to_object())
+                .unwrap_optimized()
+                .into();
+        match is_in_correct_subgroup {
+            true => Some(res),
+            false => None,
+        }
+    }
+
+    /// Multiplies a point `p0` in G2 by a scalar.
+    pub fn g2_mul(&self, p0: &G2Affine, scalar: &U256) -> G2Affine {
+        let env = self.env();
+        let bin =
+            internal::Env::bls12_381_g2_mul(env, p0.to_object(), scalar.into()).unwrap_infallible();
+        unsafe { G2Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) }
+    }
+
+    /// Performs a multi-scalar multiplication (MSM) operation in G2.
+    pub fn g2_msm(&self, vp: Vec<G2Affine>, vs: Vec<U256>) -> G2Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g2_msm(env, vp.into(), vs.into()).unwrap_infallible();
+        unsafe { G2Affine::from_bytes(BytesN::unchecked_new(env.clone(), bin)) }
+    }
+
+    /// Maps an element in the base field `Fp2` to a point in G2.
+    pub fn map_fp2_to_g2(&self, fp2: &Fp2) -> G2Affine {
```

Comment:
This can be:
- `impl From<Fp2> for G2Affine`
- `impl From<&Fp2> for G2Affine`
- `impl Fp2 { pub fn to_g2(&self) -> G2Affine { ... } }`

Reply:
Added map_to_g* methods.